---
title: "Clonealign analysis for `r params$id` with top `r 100 * params$var_quantile`% most variable genes"
output:
  html_document: 
    toc: true
    toc_float: true
  html_notebook: default
params:
  id: one
  var_quantile: 0.5
  ca_fit: clonealign_fit.rds
  cnv_prevs: cnv_prevs.csv
  input_sce: sce.rds
  input_cnv_mat: cnv.rds
  input_cnv_df: cnv.csv
  output_voom_results: output.rds
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache = FALSE
)


suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(scater)
  library(dplyr)
  library(tidyr)
  library(readr)
  library(matrixStats)
  library(pheatmap)
  library(limma)  
  library(edgeR)
})

devtools::load_all("/cellassign/clonealign/")

source("scripts/utils.R")
```


# Read in data

```{r}
ca <- readRDS(params$ca_fit)
```

```{r}
df_cnv_prev <- read_csv(params$cnv_prevs)
```

```{r}
sce <- readRDS(params$input_sce)
```

```{r}
cnv_mat <- readRDS(params$input_cnv_mat)
```

```{r}
cnv_df <- read_csv(params$input_cnv_df)
```





# Diagnostic plots

## Convergence

```{r}
qplot(seq_along(ca$elbo), ca$elbo) +
  labs(x = "Iteration", y = "ELBO")
```

## Clone assignment

```{r}
pheatmap(ca$ml_params$clone_probs)
```


```{r}
rna_clone_prev <- as_tibble(table(ca$clone)) %>% 
  dplyr::rename(clone = Var1, n_rna = n) %>% 
  dplyr::mutate(freq_rna = n_rna / sum(n_rna))
```

Compare to CNV

```{r}
df_compare_prev <- df_cnv_prev %>% 
  dplyr::rename(clone = cluster) %>% 
  dplyr::full_join(rna_clone_prev, by = "clone") %>% 
  replace_na(list(n = 0, freq = 0, n_rna = 0, freq_rna = 0))

print(df_compare_prev)
```


```{r}
ggplot(df_compare_prev, aes(x = freq, y = freq_rna)) +
  geom_abline(slope = 1, intercept = 0, colour = 'grey50') +
  # geom_point() +
  geom_text(aes(label = clone)) +
  labs(x = "Clone prevalence DNA", y = "Clone prevalence RNA") +
  theme_bw() 
```

## Expression vs copy number track plots

```{r, message = FALSE, warning = FALSE}
unique_chrs <- sort(unique(rowData(sce)$chromosome_name))

for(chr in unique_chrs) {
  tryCatch({
    print(plot_clonealign(sce[rownames(cnv_mat),], ca$clone, cnv_mat, chromosome = chr,
                    chr_str = "chromosome_name", expression_ylim = c(-.2, .2)))
  }, error = function(e) NULL)
}
```


## Correlation between genome and transcriptome

```{r}
sce_ca <- sce[rownames(cnv_mat),]

compute_correlation <- function(i) {
  y <- logcounts(sce_ca)[i,]
  x <- cnv_mat[i, ca$clone]
  c <- cor(x,y, method = "spearman")
  tibble(mean_log_expression = mean(y), correlation = c)
}

df_cor <- bind_rows(lapply(seq_len(nrow(sce_ca)), compute_correlation))
```

```{r}
ggplot(df_cor, aes(x = mean_log_expression, y=  correlation)) +
  geom_point() +
  geom_smooth()
```

```{r}
ggplot(df_cor, aes(x = correlation)) +
  geom_histogram(fill = 'grey90', colour ='black', bins = 100) +
  geom_vline(xintercept = mean(df_cor$correlation, na.rm = TRUE), linetype = 2, colour = 'darkred', size = 1) +
  theme_bw() +
  labs(x = "Genome-transcriptome correlation")

```

```{r}
print(summary(df_cor$correlation))
```


# Differential expression analysis


## Find genes in copy number altered regions


```{r}
df_clone_specific_genes <- dplyr::select(cnv_df, -entrezgene, -median_var_copy_number) %>% 
  gather(clone, copy_number, -ensembl_gene_id) %>% 
  group_by(ensembl_gene_id) %>% 
  summarise(cn_diff_between_clones = var(copy_number) > 0)

head(df_clone_specific_genes)
```

```{r}
print(table(df_clone_specific_genes$cn_diff_between_clones))
```

```{r}
non_cnv_genes <- dplyr::filter(df_clone_specific_genes, !cn_diff_between_clones) %>% 
  .$ensembl_gene_id
```


## Differential expression

Subset SCE to (i) non CNV different genes, and (ii)

```{r}
sce_de <- sce[non_cnv_genes[non_cnv_genes %in% rownames(sce)],]
```

```{r}
sce_de <- calculateQCMetrics(sce_de)

sce_de <- sce_de[rowData(sce_de)$total_counts > 200,]
```



```{r}
dge <- DGEList(counts(sce_de))
dge <- calcNormFactors(dge)

design <- model.matrix(~ ca$clone)

colnames(design) <- gsub("ca$clone", "", colnames(design), fixed = TRUE)

v <- voom(dge, design, plot = TRUE)

fit <- lmFit(v, design)
fit <- eBayes(fit)
res <- decideTests(fit)

```

Diagnostic plots:

```{r}
plotDensities(fit)
plotMA(fit)
```

## Volcano plots

```{r}
coefs <- seq_len(ncol(design) - 1) + 1

voom_results <- lapply(coefs, make_volcano, fit, sce_de, ca)
```

Display the plots:

```{r}
for(i in seq_along(voom_results)) {
  plot(voom_results[[i]]$plot)
}
```

```{r}
saveRDS(voom_results, params$output_voom_results)
```


