---
title: "Clonealign analysis for `r params$id` with top `r 100 * params$var_quantile`% most variable genes"
output:
  html_document: 
    toc: true
    toc_float: true
  html_notebook: default
params:
  id: one
  var_quantile: 0.9
  ca_fit: clonealign_fit.rds
  clone_prevs: clone_prevs.csv
  input_sce: sce.rds
  input_cnv_mat: cnv.rds
  input_cnv_df: cnv.csv
  output_voom_results: output.rds
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache = FALSE
)


suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(scater)
  library(dplyr)
  library(tidyr)
  library(tibble)
  library(readr)
  library(matrixStats)
  library(pheatmap)
  library(limma)  
  library(edgeR)
})

devtools::load_all("/cellassign/clonealign/")

# source("scripts/utils.R")
```


# Read in data

```{r}
ca <- readRDS(params$ca_fit)
```

```{r}
df_clone_prev <- read_csv(params$clone_prevs)
```

```{r}
sce <- readRDS(params$input_sce)
```

```{r}
cnv_mat <- readRDS(params$input_cnv_mat)
```

```{r}
cnv_df <- read_csv(params$input_cnv_df)
```





# Diagnostic plots

## Convergence

```{r}
qplot(seq_along(ca$elbo), ca$elbo) +
  labs(x = "Iteration", y = "ELBO")
```

## Clone assignment

```{r}
pheatmap(ca$ml_params$clone_probs)
```


```{r}
rna_clone_prev <- enframe(table(ca$clone), "clone", "n_rna") %>% 
  dplyr::filter(clone != "unassigned") %>% 
  dplyr::mutate(freq_rna = n_rna / sum(n_rna))
```

Compare to CNV

```{r}
timepoint <- sce$timepoint[1]

df_clone_prev <- dplyr::filter(df_clone_prev, time == timepoint)
```


```{r}
df_compare_prev <- df_clone_prev %>% 
  dplyr::select(-time) %>% 
  dplyr::rename(clone = cluster) %>% 
  dplyr::mutate(freq = n / sum(n)) %>%
  dplyr::rename(freq_dna = freq, n_dna = n) %>% 
  dplyr::full_join(rna_clone_prev, by = "clone") %>% 
  replace_na(list(n_dna = 0, freq_dna = 0, n_rna = 0, freq_rna = 0))

print(df_compare_prev)
```


```{r}
ggplot(df_compare_prev, aes(x = freq_dna, y = freq_rna)) +
  geom_abline(slope = 1, intercept = 0, colour = 'grey50') +
  # geom_point() +
  geom_text(aes(label = clone)) +
  labs(x = "Clone prevalence DNA", y = "Clone prevalence RNA") +
  theme_bw() 
```



## Correlation between genome and transcriptome

```{r}
df_cor <- tibble(correlation = ca$correlations)
```


```{r}
ggplot(df_cor, aes(x = correlation)) +
  geom_histogram(fill = 'grey90', colour ='black', bins = 100) +
  geom_vline(xintercept = mean(df_cor$correlation, na.rm = TRUE), linetype = 2, colour = 'darkred', size = 1) +
  theme_bw() +
  labs(x = "Genome-transcriptome correlation")

```

```{r}
print(summary(df_cor$correlation))
```




