---
title: "Preparing copy number data for input to clonealign for sample `r params$id`"
output:
  html_document: 
    toc: true
    toc_float: true
  html_notebook: default
params:
  pdx: SA535
  output_df: path_to_output.csv
  prevalence_csv: prev.csv
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache = FALSE
)


suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(TxDb.Hsapiens.UCSC.hg19.knownGene)
  library(org.Hs.eg.db)
  library(GenomicRanges)
  library(matrixStats)
  library(tidyverse)
  library(here)
})
```


# Read in data



```{r}
cnv_files <- dir(here("data", "cnv_clones", params$pdx), full.names=TRUE)

cnv_df <- map_df(cnv_files, read_csv,
                 col_types = list(chr = col_character()))
```

# Examine consistency between clones


```{r}
cnv_clone <- 
  group_by(cnv_df, chr, start, end, cluster, time) %>% 
  summarise(median_copy_number = median(copy_number)) %>% 
  ungroup()
```



```{r}
cnv_df_B <- filter(cnv_df, cluster == "B")

chr_levels <- c(as.character(1:23), "X", "Y")
cnv_df_B$chr <- factor(cnv_df_B$chr, levels = chr_levels)
cnv_df_B$copy_number_chr <- as.character(round(cnv_df_B$copy_number))
cnv_df_B$copy_number_chr[cnv_df_B$copy_number > 5] <- "6+"

cnv_cols <- c("0" = "#2166ac",
              "1" = "#92c5de", 
              "2" = "grey80", 
              "3" = "#f4a582", 
              "4" = "#d6604d",
              "5" = "#b2182b",
              "6+" = "#67001f")

ggplot(cnv_df_B, aes(x = start, y = cell_names, fill = copy_number_chr)) +
  geom_raster() +
  facet_grid(time ~ chr, scales = "free", space = "free_x", switch = "both") +
  scale_fill_manual(values = cnv_cols, name = "Copy Number") +
  labs(x = "Chromosome", y = "Clone", subtitle = "Cluster B only") +
  theme_bw() +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        strip.background = element_rect(fill = 'white'),
        legend.position = "bottom") +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) 

```






```{r}
chr_levels <- c(as.character(1:23), "X", "Y")
cnv_clone$chr <- factor(cnv_clone$chr, levels = chr_levels)
cnv_clone$copy_number_chr <- as.character(round(cnv_clone$median_copy_number))
cnv_clone$copy_number_chr[cnv_clone$median_copy_number > 5] <- "6+"

cnv_cols <- c("0" = "#2166ac",
              "1" = "#92c5de", 
              "2" = "grey80", 
              "3" = "#f4a582", 
              "4" = "#d6604d",
              "5" = "#b2182b",
              "6+" = "#67001f")

ggplot(cnv_clone, aes(x = start, y = time, fill = copy_number_chr)) +
  geom_raster() +
  facet_grid(cluster ~ chr, scales = "free", space = "free", switch = "both") +
  scale_fill_manual(values = cnv_cols, name = "Copy Number") +
  labs(x = "Chromosome", y = "Clone") +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        strip.background = element_rect(fill = 'white'),
        legend.position = "bottom") +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) 

```


```{r}
filter(cnv_df, chr == "1", cluster == "E", time == "X6") %>% 
  ggplot() +
  geom_segment(aes(x = start, xend = end, y = copy_number, yend = copy_number )) +
  facet_wrap(~ cell_names)

```


# Summarize

Group to clone level:

```{r}
clone_df <- dplyr::select(cnv_df, -cell_names, -time) %>% 
  group_by(chr, start, end, cluster) %>% 
  summarise(median_copy_number = median(copy_number),
            var_copy_number = var(copy_number)) %>% 
  ungroup()

clone_df <- replace_na(clone_df, list(var_copy_number = 0))
```

And get the minimum variance per region:

```{r}
clone_df <- group_by(clone_df, chr, start, end) %>% 
  summarise(intra_clone_var = median(var_copy_number),
            inter_clone_var = var(median_copy_number)) %>% 
  inner_join(clone_df) %>% 
  ungroup()
```

And convert to consistent name:

```{r}
df_cnv <- dplyr::select(clone_df, chr, start, end, clone = cluster, copy_number = median_copy_number, intra_clone_var, inter_clone_var)
```

Write cloneal prevalences

```{r}
prev_df <- dplyr::select(cnv_df, cell_names, cluster, time) %>% 
  distinct() %>% 
  dplyr::count(cluster, time) %>% 
  dplyr::mutate(freq = n / sum(n))

prev_df
```
```{r}
write_csv(prev_df, params$prevalence_csv)
```



# Converting region-based copy number to gene based


## Loading the  gene database and making sure chromosome names match

Next, we load the database of genes. In this example we choose the hg19 annotation, since our correspodning single-cell RNA-seq is aligned to hg19, and the copy number data (above) has been aligned to hg19. Thus we set

```{r}
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
```


We can then load the corresponding gene annotations via

```{r}
g <- genes(txdb, single.strand.genes.only=FALSE)
g
```



One issue that creates perennial worldwide angst is that our chromosome names are 1,2,3,... whereas those with gene annotation are chr1, chr2, chr3,... We fix this by changing our original data frame as follows:

```{r}
df_cnv <- dplyr::mutate(df_cnv, chr = paste0("chr", chr))
```

We can then convert this to a `GRanges` object:

```{r}
cnv_gr <- makeGRangesFromDataFrame(df_cnv, keep.extra.columns = TRUE)
cnv_gr
```

## Finding overlaps between gene and region based annotation

Then we compute the overlaps between the gene annotation and the copy number data:

```{r}
olaps <- findOverlaps(g, cnv_gr)
olaps
```

Here, the first column represents the index of each gene in `g` (that can be accessed with `queryHits(olaps)`), while the second column represents the index of each copy number region (`subjectHits(olaps)`). We can then convert this into a gene and copy number data frame:

```{r}
df_gene <- data_frame(entrezgene = names(g)[queryHits(olaps)], 
           copy_number = mcols(cnv_gr)$copy_number[subjectHits(olaps)],
           clone = mcols(cnv_gr)$clone[subjectHits(olaps)],
           intra_clone_var = mcols(cnv_gr)$intra_clone_var[subjectHits(olaps)],
           inter_clone_var = mcols(cnv_gr)$inter_clone_var[subjectHits(olaps)])
```

Next, we'd like to map on ensembl gene ids:


```{r}
entrezgene_ensembl_map <- as.list(org.Hs.egENSEMBL)
entrezgene_ensembl_map <- lapply(entrezgene_ensembl_map, `[`, 1)

df_gene <- dplyr::filter(df_gene, entrezgene %in% names(entrezgene_ensembl_map)) %>% 
  dplyr::mutate(ensembl_gene_id = unlist(entrezgene_ensembl_map[entrezgene])) %>% 
  dplyr::select(ensembl_gene_id, entrezgene, copy_number, clone, intra_clone_var, inter_clone_var) %>% 
  drop_na()

df_gene
```

We may find non-unique mappings. This can be due to genes spanning breakpoints or multi-mappings to e.g. pseudo-autosomal regions. To fix this, we retain only genes that are uniquely mapped:

```{r}
df_gene <- dplyr::count(df_gene, ensembl_gene_id) %>% 
  dplyr::filter(n == length(unique(df_gene$clone))) %>% 
  inner_join(df_gene) %>% 
  dplyr::select(-n)
```


## Creating and filtering the input for clonealign

Clonealign requires a gene by clone matrix as input, so to create this we'll use the `spread` function from `tidyr`:

```{r}
df_gene_expanded <- spread(df_gene, clone, copy_number)
head(df_gene_expanded)
```

And save this:

```{r}
write_csv(df_gene_expanded, params$output_df)
```



# Technical

```{r}
sessionInfo()
```


