---
title: "Preparing copy number data for input to clonealign for sample `r params$id`"
output:
  html_document: 
    toc: true
    toc_float: true
  html_notebook: default
params:
  pdx: SA609
  output_df: path_to_output.csv
  prevalence_csv: prev.csv
  max_copy_number_trust: 5
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache = FALSE
)


suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(TxDb.Hsapiens.UCSC.hg19.knownGene)
  library(org.Hs.eg.db)
  library(GenomicRanges)
  library(matrixStats)
  library(tidyverse)
  library(here)
})

# tensorflow::install_tensorflow(extra_packages ="tensorflow-probability", version="1.13.1")
```


# Read in data


```{r}
cnv_files <- dir(here("data", "cnv_clones", params$pdx), full.names=TRUE)

cnv_df <- map_df(cnv_files, read_csv,
                 col_types = list(chr = col_character()))
```

# Examine consistency between clones


```{r}
cnv_clone <- 
  group_by(cnv_df, chr, start, end, cluster) %>% 
  summarise(median_copy_number = median(copy_number),
            sd_copy_number = sd(copy_number)) %>% 
  ungroup() %>% 
  rename(clone = cluster)
```

Compute some useful cross-clone metrics

```{r}
cnv_clone <- group_by(cnv_clone, chr, start, end) %>% 
  summarize(between_clone_sd_copy_number = sd(median_copy_number),
            n_distinct_cnv = n_distinct(median_copy_number),
            max_copy_number = max(median_copy_number)) %>% 
  ungroup() %>% 
  inner_join(cnv_clone) %>% 
  mutate(log_ratio = log2(between_clone_sd_copy_number / sd_copy_number))
```

Make some plots:

```{r fig.width = 10}
cnv_clone$chr <- factor(cnv_clone$chr, levels = c(as.character(1:23), "X", "Y"))

cnv_clone %>% 
  ggplot(aes(x = start, xend = end, y = median_copy_number, yend = median_copy_number)) +
  geom_segment(size = 1.2) +
  facet_grid(clone ~ chr, scales = "free_x", space = "free_x")


# ggsave("../figs/deleteme.png", width = 12, height = 6)
```



```{r fig.width = 10}
cnv_clone %>% 
  ggplot(aes(x = start, xend = end, y = sd_copy_number, yend = sd_copy_number)) +
  geom_segment(size = 1.2) +
  facet_grid(clone ~ chr, scales = "free_x", space = "free_x")

# ggsave("../figs/deleteme.png", width = 12, height = 6)
```

```{r fig.width = 10}
cnv_clone %>% 
  ggplot(aes(x = start, xend = end, y = log_ratio, yend = log_ratio)) +
  geom_segment(size = 1.2) +
  facet_grid(clone ~ chr, scales = "free_x", space = "free_x")

# ggsave("../figs/deleteme.png", width = 12, height = 6)
```


```{r}
df_mlr <- group_by(cnv_clone, chr, start, end) %>% 
  summarise(median_log_ratio = median(log_ratio),
            max_copy_number = max(median_copy_number)) %>% 
  ungroup()
```

```{r}
df_mlr <- mutate(df_mlr, pass_region = median_log_ratio > 0 & max_copy_number <= params$max_copy_number_trust) %>% 
  replace_na(replace = list(pass_region = FALSE))
```


```{r fig.width = 12, fig.height = 3}
ggplot(df_mlr, aes(x = start, xend = end, y = median_log_ratio, yend = median_log_ratio, colour = pass_region)) +
  geom_segment(size = 2) +
  facet_grid(. ~ chr, scales = "free_x", space = "free_x") +
  scale_colour_brewer(palette = "Set2")
```


```{r}
cnv_clone <- inner_join(cnv_clone, df_mlr) %>% 
  ungroup()
```



```{r}
ggplot(cnv_clone, aes(x = start, xend = end, y = n_distinct_cnv, yend = n_distinct_cnv, colour = pass_region)) +
  geom_segment(size = 2) +
  facet_grid(. ~ chr, scales = "free_x", space = "free_x") +
  scale_colour_brewer(palette = "Set2")
```


# Write prevalences



Write cloneal prevalences

```{r}
prev_df <- dplyr::select(cnv_df, cell_names, cluster, time) %>% 
  distinct() %>% 
  dplyr::count(cluster, time) %>% 
  dplyr::mutate(freq = n / sum(n))

prev_df
```

```{r}
write_csv(prev_df, params$prevalence_csv)
```



# Converting region-based copy number to gene based


## Loading the  gene database and making sure chromosome names match

Next, we load the database of genes. In this example we choose the hg19 annotation, since our correspodning single-cell RNA-seq is aligned to hg19, and the copy number data (above) has been aligned to hg19. Thus we set

```{r}
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
```


We can then load the corresponding gene annotations via

```{r}
g <- genes(txdb, single.strand.genes.only=FALSE)
g
```



One issue that creates perennial worldwide angst is that our chromosome names are 1,2,3,... whereas those with gene annotation are chr1, chr2, chr3,... We fix this by changing our original data frame as follows:

```{r}
cnv_clone <- dplyr::mutate(cnv_clone, chr = paste0("chr", chr))
```

Let's add in the original co-ordinates for dbugging

```{r}
cnv_clone$.chr <- cnv_clone$chr
cnv_clone$.start <- cnv_clone$start
cnv_clone$.end <- cnv_clone$end
```


We can then convert this to a `GRanges` object:

```{r}
cnv_gr <- makeGRangesFromDataFrame(cnv_clone, keep.extra.columns = TRUE)
cnv_gr
```

## Finding overlaps between gene and region based annotation

Then we compute the overlaps between the gene annotation and the copy number data:

```{r}
olaps <- findOverlaps(g, cnv_gr)
olaps
```

Here, the first column represents the index of each gene in `g` (that can be accessed with `queryHits(olaps)`), while the second column represents the index of each copy number region (`subjectHits(olaps)`). We can then convert this into a gene and copy number data frame:

```{r}
df_gene <- tibble(entrezgene = names(g)[queryHits(olaps)])


for(n in names(mcols(cnv_gr))) {
  df_gene[[ n ]] <- mcols(cnv_gr)[[ n ]][subjectHits(olaps)]
}
```

Next, we'd like to map on ensembl gene ids:


```{r}
df_gene$ensembl_gene_id <- mapIds(org.Hs.eg.db, 
                                  keys = df_gene$entrezgene, 
                                  column="ENSEMBL", 
                                  keytype="ENTREZID", 
                                  multiVals="first")
```

We may find non-unique mappings. This can be due to genes spanning breakpoints or multi-mappings to e.g. pseudo-autosomal regions. To fix this, we retain only genes that are uniquely mapped:

```{r}
gene_counts <- dplyr::count(df_gene, entrezgene) 

df_gene <- inner_join(df_gene, gene_counts) %>% 
  rename(n_times_gene_map = n)

target_count <- unique(gene_counts$n)
target_count <- min(target_count[target_count > 0])

df_gene <- mutate(df_gene,
                  gene_span_breakpoint = n_times_gene_map > target_count)
```

Finally, do a bit of tidy up: if the gene spans a breakpoint but everything else is good, then we pass it

```{r}
df_breakpoint <- filter(df_gene, gene_span_breakpoint) %>% 
  group_by(entrezgene, clone) %>% 
  summarize(sd_breakpoint_gene = sd(median_copy_number)) %>% 
  group_by(entrezgene) %>% 
  summarize(gene_same_across_breakpoint = all(sd_breakpoint_gene == 0))
```

```{r}
genes_same_across_breakpoint <- filter(df_breakpoint, gene_same_across_breakpoint) %>% 
  .$entrezgene

# Correct those
df_gene$gene_span_breakpoint[df_gene$entrezgene %in% genes_same_across_breakpoint] <- FALSE
```

```{r}
df_gene <- mutate(df_gene, use_for_clonealign = !gene_span_breakpoint & pass_region)
```



## Creating and filtering the input for clonealign

```{r}
write_csv(df_gene, params$output_df)
```



# Technical

```{r}
sessionInfo()
```


