---
title: "Clonealign analysis for `r params$id` with top `r 100 * params$var_quantile`% most variable genes"
output:
  html_document: 
    toc: true
    toc_float: true
  html_notebook: default
params:
  id: one
  var_quantile: 0.9
  ca_fit: clonealign_fit.rds
  clone_prevs: clone_prevs.csv
  input_sce: sce.rds
  input_cnv_mat: cnv.rds
  input_cnv_df: cnv.csv
  output_voom_results: output.rds
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache = FALSE
)


suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(scater)
  library(dplyr)
  library(tidyr)
  library(tibble)
  library(readr)
  library(matrixStats)
  library(pheatmap)
  library(limma)  
  library(edgeR)
  library(cowplot)
  library(here)
})
```


# Read in data


```{r}
sces <- list(
  X3 = readRDS(here("/data/sces/sce-qc_SA609X3XB01584.rds")),
  X6 = readRDS(here("data/sces/sce-qc_SA609X6XB01899.rds")),
  X10 = readRDS(here("data/sces/sce-qc_SA609X10XB02454.rds"))
)
```

```{r}
timepoints <- names(sces)
```


```{r}
clonealign_fits <- list(
  X3 = readRDS("../data/clonealign_fits/SA609X3XB01584/clonealign-SA609X3XB01584-var_0.3.rds"),
  X6 = readRDS("../data/clonealign_fits/SA609X6XB01899/clonealign-SA609X6XB01899-var_0.3.rds"),
  X10 = readRDS("../data/clonealign_fits/SA609X10XB02454/clonealign-SA609X10XB02454-var_0.3.rds")
)
```

```{r}
for(timepoint in timepoints) {
  sces[[ timepoint ]]$clone <- clonealign_fits[[ timepoint ]]$clone
}
```

Load in CNV data here

```{r}
cnv <- read_csv(here("data/processed_cnv/cnv_SA609.csv"))
```


```{r}
dlp_clone_prev <- read_csv(here("data/processed_cnv/clone_prevalence_SA609.csv"))
```


# Visualizations

Colour scheme:

```{r}
sa609_cols <- c("A"="#66C2A5",
                "C"="#FC8D62",
                "B"="#8DA0CB",
                "D"="#E78AC3",
                "E"="#A6D854")

scale_clone_sa609 <- function(...) {
  scale_colour_manual(values = sa609_cols, name = "Clone", ...)
}
```


```{r}
sces <- lapply(sces, runUMAP)
```

```{r}
umap_dims <- lapply(sces, function(sce) reducedDims(sce)[['UMAP']])
umap_dims <- do.call('rbind', umap_dims)
clones <- unlist(sapply(sces, function(sce) sce$clone))
cell_timepoints <- unlist(sapply(sces, function(sce) sce$timepoint))

df_umap <- as_tibble(umap_dims) %>% 
  dplyr::mutate(clone = clones,
         timepoint = factor(cell_timepoints, levels = timepoints))

```

```{r}
ggplot(df_umap, aes(x = V1, y = V2, colour = clone)) +
  geom_point() +
  facet_wrap(~ timepoint, nrow = 1) +
  scale_colour_brewer(palette = "Spectral")
```


```{r}
dlp_clone_prev <- dlp_clone_prev %>% 
  group_by(time) %>% 
  mutate(freq = n / sum(n))
```


```{r}
clones_counted <- count(df_umap, clone, timepoint)

clones_counted <- group_by(clones_counted, timepoint) %>%
  mutate(prev = n / sum(n)) %>% 
  ungroup()

# dirty trick to get 0s in missing entries
clones_counted <- select(clones_counted, -n) %>%
  spread(timepoint, prev, fill = 0) %>%
  gather(timepoint, prev, -clone) %>% 
  mutate(timepoint = factor(timepoint, levels = timepoints))

clones_counted <- rename(dlp_clone_prev, clone = cluster, timepoint = time, prev = freq) %>% 
  select(-n) %>% 
  right_join(clones_counted, suffix = c("_dlp", "_rna"), by = c("clone", "timepoint")) %>% 
  replace_na(list(prev_dlp = 0))

clones_counted <- ungroup(clones_counted)

clones_counted <- rename(clones_counted, DLP = prev_dlp, RNA = prev_rna) %>% 
  gather(modality, prevalence, -clone, -timepoint) %>% 
  mutate(timepoint = factor(timepoint, levels = timepoints))

ggplot(clones_counted, aes(x = timepoint, y = prevalence, 
                           colour = clone, group = clone)) +
  geom_line(size = 1.5) +
  geom_point(size = 3) +
  labs(x = "Timepoint", y = "Clonal prevalence") +
  scale_clone_sa609() +
  facet_wrap(~ modality)
```

```{r}
ggsave("../figs/clonal-prevalence-sa609.png", width = 10, height = 4)
```



# Differential expression

Useful functions:

```{r}
annotate_cnv_genes <- function(sce, cnv_genes, no_cnv_genes) {
  rowData(sce)$cnv_gene <- "unknown"
  rowData(sce)$cnv_gene[rownames(sce) %in% cnv_genes] <- "cnv"
  rowData(sce)$cnv_gene[rownames(sce) %in% no_cnv_genes] <- "no_cnv"
  sce
}
```


## Confirmatory

### A vs C

```{r}
cnv_a_vs_c <- select(cnv, one_of(c("ensembl_gene_id", "entrezgene", "intra_clone_var", "inter_clone_var", "A", "C")))

a_vs_c_cnv_gene <- filter(cnv_a_vs_c, A != C) %>% .$ensembl_gene_id
a_vs_c_no_cnv_gene <- filter(cnv_a_vs_c, A == C) %>% .$ensembl_gene_id
```

#### A vs C at X3

```{r}
sceX3 <- sces$X3
sceX3 <- annotate_cnv_genes(sceX3, a_vs_c_cnv_gene, a_vs_c_no_cnv_gene)

sceX3_de <- sceX3[rowData(sceX3)$cnv_gene == "no_cnv", sceX3$clone %in% c("A", "C")]
```

```{r}
sceX3_de <- sceX3[rowSums(as.matrix(counts(sceX3_de))) > 50,]
```

```{r}
dge <- DGEList(counts(sceX3_de))
dge <- calcNormFactors(dge)

design <- model.matrix(~ sceX3_de$clone)

colnames(design) <- gsub("sceX3_de$clone", "", colnames(design), fixed = TRUE)

v <- voom(dge, design, plot = TRUE)

fit <- lmFit(v, design)
fit <- eBayes(fit)
res <- decideTests(fit)
```


#### A vs C at X6


### D vs E

#### D vs E at X6

#### D vs E at X10



## X6 DE

# Diagnostic plots

## Convergence

```{r}
qplot(seq_along(ca$elbo), ca$elbo) +
  labs(x = "Iteration", y = "ELBO")
```

## Clone assignment

```{r}
pheatmap(ca$ml_params$clone_probs)
```


```{r}
rna_clone_prev <- enframe(table(ca$clone), "clone", "n_rna") %>% 
  dplyr::mutate(freq_rna = n_rna / sum(n_rna))
```

Compare to CNV

```{r}
timepoint <- sce$timepoint[1]

df_clone_prev <- filter(df_clone_prev, time == timepoint)
```


```{r}
df_compare_prev <- df_clone_prev %>% 
  dplyr::select(-time) %>% 
  dplyr::rename(clone = cluster) %>% 
  mutate(freq = n / sum(n)) %>%
  rename(freq_dna = freq, n_dna = n) %>% 
  dplyr::full_join(rna_clone_prev, by = "clone") %>% 
  replace_na(list(n_dna = 0, freq_dna = 0, n_rna = 0, freq_rna = 0))

print(df_compare_prev)
```


```{r}
ggplot(df_compare_prev, aes(x = freq_dna, y = freq_rna)) +
  geom_abline(slope = 1, intercept = 0, colour = 'grey50') +
  # geom_point() +
  geom_text(aes(label = clone)) +
  labs(x = "Clone prevalence DNA", y = "Clone prevalence RNA") +
  theme_bw() 
```

## Expression vs copy number track plots

```{r, message = FALSE, warning = FALSE}
unique_chrs <- sort(unique(rowData(sce)$chromosome_name))

for(chr in unique_chrs) {
  tryCatch({
    print(plot_clonealign(sce[rownames(cnv_mat),], ca$clone, cnv_mat, chromosome = chr,
                    chr_str = "chromosome_name", expression_ylim = c(-.2, .2)))
  }, error = function(e) NULL)
}
```


## Correlation between genome and transcriptome

```{r}
sce_ca <- sce[rownames(cnv_mat),]

compute_correlation <- function(i) {
  y <- logcounts(sce_ca)[i,]
  x <- cnv_mat[i, ca$clone]
  c <- cor(x,y, method = "spearman")
  tibble(mean_log_expression = mean(y), correlation = c)
}

df_cor <- bind_rows(lapply(seq_len(nrow(sce_ca)), compute_correlation))
```

```{r}
ggplot(df_cor, aes(x = mean_log_expression, y=  correlation)) +
  geom_point() +
  geom_smooth()
```

```{r}
ggplot(df_cor, aes(x = correlation)) +
  geom_histogram(fill = 'grey90', colour ='black', bins = 100) +
  geom_vline(xintercept = mean(df_cor$correlation, na.rm = TRUE), linetype = 2, colour = 'darkred', size = 1) +
  theme_bw() +
  labs(x = "Genome-transcriptome correlation")

```

```{r}
print(summary(df_cor$correlation))
```


# Differential expression analysis


## Find genes in copy number altered regions


```{r}
df_clone_specific_genes <- dplyr::select(cnv_df, -entrezgene, -intra_clone_var, -inter_clone_var) %>% 
  gather(clone, copy_number, -ensembl_gene_id) %>% 
  group_by(ensembl_gene_id) %>% 
  summarise(cn_diff_between_clones = var(copy_number) > 0)

head(df_clone_specific_genes)
```

```{r}
print(table(df_clone_specific_genes$cn_diff_between_clones))
```

```{r}
non_cnv_genes <- dplyr::filter(df_clone_specific_genes, !cn_diff_between_clones) %>% 
  .$ensembl_gene_id
```


## Differential expression

Subset SCE to (i) non CNV different genes, and (ii)

```{r}
sce_de <- sce[non_cnv_genes[non_cnv_genes %in% rownames(sce)],]
```

```{r}
sce_de <- calculateQCMetrics(sce_de)

sce_de <- sce_de[rowData(sce_de)$total_counts > 200,]
```



```{r}
dge <- DGEList(counts(sce_de))
dge <- calcNormFactors(dge)

design <- model.matrix(~ ca$clone)

colnames(design) <- gsub("ca$clone", "", colnames(design), fixed = TRUE)

v <- voom(dge, design, plot = TRUE)

fit <- lmFit(v, design)
fit <- eBayes(fit)
res <- decideTests(fit)

```

Diagnostic plots:

```{r}
plotDensities(fit)
plotMA(fit)
```

## Volcano plots

```{r}
coefs <- seq_len(ncol(design) - 1) + 1

voom_results <- lapply(coefs, make_volcano, fit, sce_de, ca)
```

Display the plots:

```{r}
for(i in seq_along(voom_results)) {
  plot(voom_results[[i]]$plot)
}
```

```{r}
saveRDS(voom_results, params$output_voom_results)
```


